system/Database/BaseBuilder.php

```diff
@@ -2061,6 +2079,76 @@ public function onConstraint($set)
         return $this;
     }
 
+    /**
+     * Sets data source as a query for insert/update/upsert
+     *
+     * @param BaseBuilder|RawSql $query
+     */
+    public function fromQuery($query): BaseBuilder
+    {
+        if ($query instanceof BaseBuilder) {
+            $query = $query->getCompiledSelect();
+        } elseif ($query instanceof RawSql) {
+            $query = $query->__toString();
+        }
+
+        if (is_string($query)) {
+            $this->QBOptions['fromQuery'] = $query;
+            $this->QBKeys                 = $this->fieldsFromQuery($query);
+            $this->QBSet                  = [];
+
+            foreach ($this->QBKeys as $key => $value) {
+                $this->QBKeys[$key] = $this->db->escapeChar . $value . $this->db->escapeChar;
+            }
+        }
+
+        return $this;
+    }
+
+    /**
+     * Gets column names from a select query
+     */
+    protected function fieldsFromQuery(string $sql): array
+    {
+        $sql = preg_replace('/\\(([^()]*+|(?R))*\\)/', '', $sql); // remove everything in parenthesis - removes "FROM" and commas
+        $d   = ['`', "'", '"']; // delimeters - delimits identifiers
```

Comment:
Please replace the variable names with meaningful ones.
We don't accept single letter variable name except for loop counter.
> Code should be written to minimize the time it would take for someone else to understand it.

```suggestion
        $identifierDelimeters = ['`', "'", '"'];
```

Comments for what it is are generally not good practice. 
Making the code more readable is better.